### ISSUE TITLE
Superclass constructor calls subclass' methods before initializing its attributes

### ISSUE THREAD

**Issue by IgorSamurovic** [wurstscript/WurstScript#545]
According to some logic, what this script should do is print out "SOME VALUE" when you run the map. Instead, what happens is, it prints (null). The reasoning behind this is that, when we call super() from the subclass, the superclass constructor DOES call the subclass' onCreate method, BUT it does so BEFORE INITIALIZING the subclass' attributes. This is an extremely easily reproducible error.

I can see how this was implemented - wurst is coded to finish calling the superclass constructor before initializing any of the subclass' attributes, so fixing this issue would require some detection for whether any subclass methods were called and then initialize the attributes.

```
package Bug

class Super

    function onCreate()

    construct()
        onCreate()

class Sub extends Super

    private let someVar = "SOME VALUE"

    override function onCreate()
        print(someVar)

    construct()
        super()

init
    new Sub()
```
This is not much of an issue in this particular case since I could just initialize this value in the onCreate() but if you learn to depend on initializing attributes the way it is done in this script, in long chains of inheritances this is going to cause phantom bugs (well, especially for those unaware of the issue).

It's also worth noting that this issue persists when the superclass is abstract, and the method called is also abstract, that's how it was in my implementation, too.

**Comment by Frotty** [comment]
While I agree that it's not ideal, this is intended behavior which also throws warnings in other languages like Java. I believe there is no easy/decisive fix for this that covers all cases.

I would argue to use the constructors for what they are worth and add a warning for calling overloaded functions in them.

**Comment by IgorSamurovic** [comment]
Yeah, that's fair, while not a top priority fix, having a warning for trying this would be nice. I was more so confused because this works (to a greater extent) in Java, though it also gives you a warning.

So I would agree that detecting this and giving a warning would be the fix, instead of making code like this work with the possibility of breaking other things.

**Comment by peq** [comment]
Detecting this is not so easy and coming up with a good design is not so easy as well. Java's initialization logic might be a bit more complex, but it still has some corner cases.

One possible change is discussed in https://github.com/peq/WurstScript/issues/158, which would make initialization more flexible.

**Comment by peq** [comment]
Closing, because issue is addressed in #158
